crates/aggkit-prover/src/rpc.rs

```diff
@@ -73,70 +75,117 @@ impl AggchainProofGrpcService for GrpcService {
+        let mut context = aggchain_proof_inputs.context();
+
+        let proof_request = AggchainProofServiceRequest::Normal(aggchain_proof_inputs);
+
+        let mut service = self.service.clone();
+
+        let service = service
+            .ready()
+            .await
+            .map_err(|_| Status::internal("Unable to get the service"))?;
+
+        match service.call(proof_request).await {
+            Ok(response) => {
+                context.insert(
+                    "public_values".to_owned(),
+                    Bytes::from(
+                        bincode::serialize(&response.public_values)
+                            .unwrap_or_else(|_| b"bincode serialization failed".to_vec()),
+                    ),
+                );
+                context.insert(
+                    "local_exit_root_hash".to_owned(),
+                    Bytes::from(response.local_exit_root_hash.as_bytes().to_vec()),
+                );
+                context.insert(
+                    "end_block".to_owned(),
+                    Bytes::from(response.end_block.to_be_bytes().to_vec()),
+                );
+                Ok(Response::new(GenerateAggchainProofResponse {
+                    aggchain_proof: Some(AggchainProof {
+                        aggchain_params: Some(response.aggchain_params.into()),
+                        context,
+                        proof: Some(agglayer_interop::grpc::v1::aggchain_proof::Proof::Sp1Stark(
+                            Sp1StarkProof {
+                                version: SP1_CIRCUIT_VERSION.to_string(),
+                                proof: response.proof.into(),
+                                vkey: response.vkey.into(),
+                            },
+                        )),
+                    }),
+                    last_proven_block: response.last_proven_block,
+                    end_block: response.end_block,
+                    local_exit_root_hash: Some(response.local_exit_root_hash.into()),
+                    custom_chain_data: response.custom_chain_data.into(),
+                }))
+            }
+            Err(e) => {
+                let e = match e.downcast::<aggchain_proof_service::Error>() {
+                    Ok(e) => e,
+                    Err(e) => return Err(Status::internal(e.to_string())),
+                };
+                match *e {
+                    aggchain_proof_service::Error::ProposerServiceError(
+                        proposer_service::Error::Client(
+                            proposer_client::Error::AggProofRequestFailed(
+                                jsonrpsee::core::client::error::Error::Call(e),
+                            ),
+                        ),
+                    ) if e.code() == -2000 => Err(Status::unavailable(
+                        "Proposer service has not built any proof yet",
+                    )),
+                    e => Err(Status::internal(e.to_string())),
+                }
+            }
```

Comment:
Seems like fully duplicated code block for two kind of requests, should we move error processing to a helper function?

Reply:
That can be handled when we properly define the aggkit prover gRPC errors